File: cpp/0897.0_Increasing_Order_Search_Tree.cpp
/*
Runtime: 6 ms, faster than 20.69% of C++ online submissions for Increasing Order Search Tree.
Memory Usage: 7.6 MB, less than 77.99% of C++ online submissions for Increasing Order Search Tree.
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* hair = new TreeNode();
    TreeNode* increasingBST(TreeNode* root) {
        TreeNode* cur = hair;
        function<void(TreeNode*)> inorder = [&] (TreeNode* node) -> void {
            if (nullptr == node) return;
            inorder(node->left);
            node->left = nullptr;
            cur->right = node;
            cur = node;
            inorder(node->right);
        };
        inorder(root);
        return hair->right;
    }
};

File: java/0897.0_Increasing_Order_Search_Tree.java
/*
T: O(N)
S: O(H), for recursion stack

Runtime: 0 ms, faster than 100.00% of Java online submissions for Increasing Order Search Tree.
Memory Usage: 40 MB, less than 84.76% of Java online submissions for Increasing Order Search Tree.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode cur;
    public TreeNode increasingBST(TreeNode root) {
        TreeNode hair = new TreeNode(0);
        cur = hair;
        inorder(root);
        return hair.right;
    }
    
    public void inorder(TreeNode node) {
        if (node == null) return;
        inorder(node.left);
        node.left = null;
        cur.right = node;
        cur = node;
        inorder(node.right);
    }
}


/*
Runtime: 0 ms, faster than 100.00% of Java online submissions for Increasing Order Search Tree.
Memory Usage: 39.8 MB, less than 88.54% of Java online submissions for Increasing Order Search Tree.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode increasingBST(TreeNode root) {
        List<Integer> nums = new ArrayList();
        inorder(root, nums);
        TreeNode hair = new TreeNode(0), cur = hair;
        for (var x: nums) {
            cur.right = new TreeNode(x);
            cur = cur.right;
        }
        return hair.right;
    }
    
    public void inorder(TreeNode node, List<Integer> nums) {
        if (node == null) return;
        inorder(node.left, nums);
        nums.add(node.val);
        inorder(node.right, nums);
    }
}

